buku

```diff
@@ -3276,6 +3274,36 @@ n: don't add parent folder as tag
 class ExtendedArgumentParser(argparse.ArgumentParser):
     """Extend classic argument parser."""
 
+    def __init__(self, *args, **kwargs):
+        self._nodefaults, self._unset = None, object()
+        super().__init__(*args, **kwargs)
+        self._nodefaults = argparse.ArgumentParser(*args, **kwargs)
+
+    def _add_argument(self, old_add_arg, nodefaults, *args, **kwargs):
+        old_add_arg(*args, **kwargs)
+        kwargs = dict(kwargs)
+        kwargs.pop('type', None)
+        kwargs.pop('choices', None)
+        kwargs['default'] = self._unset
+        nodefaults and nodefaults.add_argument(*args, **kwargs)
+
+    def add_argument(self, *args, **kwargs):
+        self._add_argument(super().add_argument, self._nodefaults, *args, **kwargs)
+
+    def add_argument_group(self, *args, **kwargs):
+        group = super().add_argument_group(*args, **kwargs)
+        nodefaults = self._nodefaults and self._nodefaults.add_argument_group(*args, **kwargs)
+        old_add_arg = group.add_argument
+        group.add_argument = lambda *a, **kw: self._add_argument(old_add_arg, nodefaults, *a, **kw)
+        return group
+
+    def parse_args(self, *args, **kwargs):
+        result = super().parse_args(*args, **kwargs)
+        nodefaults = self._nodefaults.parse_args(*args, **kwargs)
+        params = {k for k in dir(nodefaults) if not k.startswith('_')}
+        setattr(result, '_passed', {k for k in params if getattr(nodefaults, k) != self._unset})
+        return result
+
```

Comment:
…Despite being intended for CLI argument parsing, `argparse` is missing the functionality to determine which arguments were actually detected in the input. This workaround implements it (with some use of monkeypatching :melting_face:)